/*! Homology.js, version 0.2.0.  Developed by Eric Weitz.  https://github.com/eweitz/homology.  Public domain (CC0 1.0). */
!function(e,t){if("object"==typeof exports&&"object"==typeof module)module.exports=t();else if("function"==typeof define&&define.amd)define([],t);else{var o=t();for(var r in o)("object"==typeof exports?exports:e)[r]=o[r]}}(window,function(){return function(e){var t={};function o(r){if(t[r])return t[r].exports;var n=t[r]={i:r,l:!1,exports:{}};return e[r].call(n.exports,n,n.exports,o),n.l=!0,n.exports}return o.m=e,o.c=t,o.d=function(e,t,r){o.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},o.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},o.t=function(e,t){if(1&t&&(e=o(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(o.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var n in e)o.d(r,n,function(t){return e[t]}.bind(null,n));return r},o.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return o.d(t,"a",t),t},o.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},o.p="/dist",o(o.s=0)}([function(e,t,o){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r,n=o(1),a=(r=n)&&r.__esModule?r:{default:r};window.fetchOrthologs=a.default,t.default=a.default},function(e,t,o){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r=a(o(2)),n=a(o(3));function a(e){return e&&e.__esModule?e:{default:e}}t.default=async function(e,t,o,a="oma"){return"oma"===a?await(0,r.default)(e,t,o):"orthodb"===a?await(0,n.default)(e,t,o):void 0}},function(e,t,o){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r="https://omabrowser.org/api";function n(e,t=null,o=null,r=null,n=null){var a,i,u;throw a={geneNotFound:'Gene "'+o+'" not found in '+r,orthologsNotFound:'Orthologs not found for gene "'+o+'"',orthologsNotFoundInTarget:'Orthologs not found for gene "'+o+'" in target organism '+n},u=t?`<small>${t.message}</small>`:"",i=`<span id="error-container">\n    <b>Error</b>: ${a[e]}.<br/>\n    ${u}\n    </span>`.replace("Error: ",""),Error(i)}t.default=async function(e,t,o){var a,i,u,s,c,l,f,d;e=e.toLowerCase();try{a=await async function(e,t){var o,r,n,a,i,u,s,c;for(o="https://www.uniprot.org/uniprot/?format=tab&"+(i="columns=id,entry name,genes")+"&sort=score",r="&query=gene:"+e+"+AND+organism:"+t,n=await fetch(o+r),a=(await n.text()).split("\n").slice(1,-1),u=0;u<a.length;u++)if(s=(i=a[u].split("\t"))[2].split(" ").map(e=>e.toLowerCase()),c=i[0],s.includes(e))return c;throw Error('No Uniprot entry found for gene name "'+e+'" in organism "'+t+'"')}(e,t),i=await async function(e){var t,o,n;if(t=r+"/protein/"+e+"/",o=await fetch(t),n=await o.json(),404===o.status)throw Error('OMA protein not found for Uniprot protein "'+e+'".');return n}(a)}catch(f){n("geneNotFound",f,e,t,[o])}try{u=await async function(e){var t,o,n;if(t=r+"/protein/"+e+"/orthologs/",o=await fetch(t),n=await o.json(),404===o.status)throw Error('OMA orthologs not found for Uniprot protein "'+e+'".');return n}(a)}catch(f){n("orthologsNotFound",f)}return d=o.map(e=>(function(e){var t,o;return"homo sapiens"===e?"HUMAN":"mus musculus"===e?"MOUSE":([t,o]=e.split(" "),(t.slice(0,3)+o.slice(0,2)).toUpperCase())})(e)),0===(l=u.filter(e=>(s=e.omaid,c=s.slice(0,5),d.includes(c)))).length&&n("orthologsNotFoundInTarget",void 0),l.unshift(i),l=l.map(e=>e.chromosome+":"+e.locus.start+"-"+e.locus.end)}},function(e,t,o){"use strict";Object.defineProperty(t,"__esModule",{value:!0});var r="https://homology-api.firebaseapp.com/orthodb",n="https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=gene&version=2.0&retmode=json&api_key=e7ce8adecd69d0457df7ec2ccbb704c4e709";async function a(e){var t=await fetch(r+e);return(await t.json()).data}async function i(e){var t=e.gene_id.param,o=(await a("/ogdetails?id="+t)).entrez[0].id;return await async function(e){var t,o,r;return t=await fetch(n+"&id="+e),(r=(o=(await t.json()).result)[o.uids[0]].genomicinfo[0]).chrloc+":"+r.chrstart+"-"+r.chrstop}(o)}async function u(e,t,o,r){var n,i=[],u=!1;return(await a(`/orthologs?id=${e}&species=all`)).forEach(e=>{var a=e.organism.name.toLowerCase();o===a&&(n=e,e.genes.forEach(e=>{var o=e.gene_id.id.toLowerCase();t.toLowerCase()===o&&(u=!0)})),r.includes(a)&&i.push(e)}),[n,i,u]}t.default=async function(e,t,o){var r,n,s,c,l=!1,f=[];for(r=await a("/search?query="+e+"&&level=2759&species=9606"),n=0;n<r.length&&(s=r[n],[c,f,l]=await u(s,e,t,o),!l);n++);if(void 0===c)throw Error(`Ortholog not found for "${e}" in source organism "${t}"`);var d,p=await i(c.genes[0]);return(d=(d=await Promise.all(f.map(async e=>await Promise.all(e.genes.map(async e=>i(e))))))[0]).unshift(p),d}}])});
//# sourceMappingURL=homology.min.js.map